docs/architecture/adr-001-node-key-refactoring.md

```diff
@@ -0,0 +1,104 @@
+# ADR ADR-001: Node Key Refactoring
+
+## Changelog
+
+- 2022-10-31: First draft
+
+## Status
+
+Proposed
+
+## Context
+
+The original key format of IAVL nodes is a hash of the node. It does not take advantage of data locality on disk. Nodes are stored in random locations on disk due to the random hash value, so it needs to scan the disk to find the corresponding node which can be very inefficient.
+
+The `orphans` are used to manage node removal in the current design and allow deletion of removed nodes for the specific version from the disk through the `DeleteVersion` API. It needs to track every time when updating the tree and also requires extra storage to store `orphans`. But there are only 2 use cases for `DeleteVersion`:
+
+1. Rollback of the tree to a previous version
+2. Remove unnecessary old nodes
+
+## Decision
+
+- Use the version and the sequenced integer ID as a node key like `bigendian(version) | bigendian(nonce)` format. 
+- Remove the `version` field from node body writes.
+- Remove the `leftHash` and `rightHash` fields, and instead store `hash` field.
+- Remove the `orphans` from the tree.
+
+New node structure
+
+```go
+type NodeKey struct {
+    version int64
+    nonce int32
```

Comment:
I have a concern about the proposed nonce-approach still causing superfluous compactions and merging that are currently the biggest bottleneck here.

By writing key-value pairs that are already lexicographically sorted, we reduce superfluous merging and compactions overhead.

The main benefit of the initially proposed approach with a path encoded as a key is that writes would be sequential, reducing the superfluous merging.

While the "nonce approach" may show an improvement, it will still have redundant compactions due to nodes intra-version not being close to each other.

This might also affect iterations and queries.

Reply:
With a monotonic increasing nonce, the inserted keys are also lexicographically sorted?

Reply:
For example:
- version 1 create 3 nodes, they are inserted with keys: `(1, 0)`, `(1, 1)`, `(1, 2)`.
- version 2 create 2 nodes, they are inserted with keys: `(2, 0)`, `(2, 1)`
- etc.

Reply:
I agree with @yihuang, it is already sorted by the version, and also we will make the nonce sorted in `SaveBranch`.